mmgen/models/architectures/biggan/biggan_snmodule.py

```diff
@@ -0,0 +1,257 @@
+            The length of list equals to the simulation times.
+        update (bool, optional): Whether update left singular
+            vector. Defaults to True.
+        eps (float, optional): Vector Normalization epsilon.
+            Defaults to 1e-12.
+
+    Returns:
+        tuple[list[tensor.Tensor]]: Tuple consist of three lists
+            which contain singular values, left singular
+            vector and right singular vector respectively.
+    """
+    us, vs, svs = [], [], []
+    for i, u in enumerate(u_list):
+        with torch.no_grad():
+            v = torch.matmul(u, weight)
+            v = F.normalize(gram_schmidt(v, vs), eps=eps)
+            vs += [v]
+            u = torch.matmul(v, weight.t())
+            u = F.normalize(gram_schmidt(u, us), eps=eps)
+            us += [u]
+            if update:
+                u_list[i][:] = u
+        svs += [
+            torch.squeeze(torch.matmul(torch.matmul(v, weight.t()), u.t()))
+        ]
+    return svs, us, vs
+
+
+class SpectralNorm(object):
+    """Spectral normalization base class.
+
+    Args:
+        num_svs (int): Number of singular values.
+        num_itrs (int): Number of power iterations per step.
+        num_outputs (int): Number of output channels.
+        transpose (bool, optional): If set to `True`, weight
+            matrix will be transposed before power iteration.
+            Defaults to False.
+        eps (float, optional): Vector Normalization epsilon for
+            avoiding divide by zero. Defaults to 1e-12.
+    """
+
+    def __init__(self,
+                 num_svs,
+                 num_itrs,
+                 num_outputs,
+                 transpose=False,
+                 eps=1e-12):
+        self.num_itrs = num_itrs
+        self.num_svs = num_svs
+        self.transpose = transpose
+        self.eps = eps
+        # Register a singular vector for each sv
+        for i in range(self.num_svs):
+            self.register_buffer('u%d' % i, torch.randn(1, num_outputs))
+            self.register_buffer('sv%d' % i, torch.ones(1))
+
+    @property
+    def u(self):
+        """Get left singular vectors."""
+        return [getattr(self, 'u%d' % i) for i in range(self.num_svs)]
+
+    @property
+    def sv(self):
+        """Get singular values."""
+        return [getattr(self, 'sv%d' % i) for i in range(self.num_svs)]
+
+    def sn_weight(self):
+        """Compute the spectrally-normalized weight."""
+        W_mat = self.weight.view(self.weight.size(0), -1)
+        if self.transpose:
+            W_mat = W_mat.t()
+        # Apply num_itrs power iterations
+        for _ in range(self.num_itrs):
+            svs, us, vs = power_iteration(
+                W_mat, self.u, update=self.training, eps=self.eps)
+        # Update the svs
+        if self.training:
+            with torch.no_grad():
+                for i, sv in enumerate(svs):
+                    self.sv[i][:] = sv
+        return self.weight / svs[0]
```

Comment:
Please leave a comment here to specify why we just use `svs[0]` rather than another element in `svs`.

Reply:
I just change `svs[0]` to `svs[-1]` instead.